File: 410-split-array-largest-sum/410-split-array-largest-sum.java
class Solution {
    public int splitArray(int[] nums, int m) {
        int sum = 0, max = Integer.MIN_VALUE;
        for(int num : nums) {
            sum += num;
            max = Math.max(max, num);
        }
        int l = max, r = sum;
        while( l < r) {
            int mid = l + (r - l) / 2;
            if(getSplits(nums, mid) <= m)
                r = mid;
            else 
                l = mid + 1;
        }
        return l;
    }
    public int getSplits(int[] nums, int mid) {
        int count = 1;
        int sum = 0;
        for(int num : nums) {
            if(sum + num <= mid)
                sum += num;
            else {
                sum = num;
                count++;
            }
        }
        return count;
    }
}

File: 410-split-array-largest-sum/410-split-array-largest-sum.py
class Solution:
    def splitArray(self, nums: List[int], m: int) -> int:
        total, max_num = 0, nums[0]
        for num in nums:
            max_num = max(max_num, num)
            total += num
        l, r=  max_num, total
        
        def countSubarray(nums, num_sum):
            cnt,  curr_sum= 1, 0
            for num in nums:
                if curr_sum + num <= num_sum:
                    curr_sum += num
                else:
                    cnt += 1
                    curr_sum = num
            return cnt
        
        while l < r:
            mid = (l + r) // 2
            if countSubarray(nums, mid) <= m:
                r = mid
            else:
                l = mid + 1
        return l